### ISSUE TITLE
0.7.x hangs on Google Cloud components using Python 3.8

### ISSUE THREAD

**Issue by yiga2** [dask/gcsfs#320]
As reported in https://github.com/dask/gcsfs/issues/278#issuecomment-685125685 and later comments, 0.7.x runs into issues under some conditions on GCP. 

Symptoms are:

- `gcsfs.GCSFileSystem(token='cloud')` or `gcsfs.GCSFileSystem()` executes w/o error - doesn't mean correctly instantiated
- following `ls(), exists() or glob()` (confirmed methods so far) run forever, no error raised
- experienced on GCF, GKE and Composer

Workaround: stick to `0.6.2`

EDIT: important finding: latest `gcsfs` works fine on GCP with Python 3.7. This could indicate a library conflict.

**Comment by rm3195** [comment]
In my case python version is the same (3.6.10) in 2 Composer environments, but GCC version differs (gcsfs works well with `GCC 5.4.0 20160609`, hangs with `GCC 7.5.0`). Not sure if that could be the cause though.

**Comment by martindurant** [comment]
Can you please list the versions of aiohttp, fsspec, python, google sdk and anything else that looks like it might be involved? The env is installed using pip, right? It may have something to do with the new version of pip.

@rm3195 - I can't see how GCC would be involved, unless I am very wrong.

**Comment by martindurant** [comment]
can you please call `gcs.maybe_refresh()` immediately after instantiation, to see if anything happens? It should populate `gcs.heads`/`gcs.credentials`. It may be worthwhile explicitly calling `gcs._connect_cloud()` too; or *if* you have a JSON credentials file, perhaps `gcs._connect_token/service()`.

Referring to the code in https://github.com/dask/gcsfs/issues/278#issuecomment-749249318 , perhaps I would try moving `gcsfs.GCSFileSystem()` *inside* the function, just in case it needs the context to already exist when it gets called.

**Comment by rm3195** [comment]
Moving `gcsfs.GCSFileSystem()` inside the function actually works, thanks! Although interesting why it worked another way before...

**Comment by yiga2** [comment]
Confirming instantiating in the function vs. global works too in GCF. 
Fine with me going forward but as @rm3195 indicates, why different with 3.8 vs. 3.7, or 0.6.2 vs. 0.7.x remains a mystery.

Note that this did not work:
```
import gcsfs

gcs = gcsfs.GCSFileSystem()
gcs.maybe_refresh()
gcs._connect_cloud()


def hello_gcs(event, context):

    file_name = event['name']
    input_bucket = event['bucket']

    file_path = f'{input_bucket}/{file_name}'
    if gcs.exists(file_path):
        print('Confirmed exists: ' + file_path)
```

**Comment by martindurant** [comment]
I would also try inverting the order
```
gcs = gcsfs.GCSFileSystem()
gcs._connect_cloud()
gcs.maybe_refresh()
```
but I suspect that 
- creating the context (before calling the function, and after import, apparently) in the google lib creates additional information relevant to auth
- the timing of creating the credentials object versus refreshing the token is not the same as it used to be, but I'm not sure what the difference is.

Thanks for investigating and bearing with us...

**Comment by martindurant** [comment]
That still doesn't answer why the calls hang indefinitely - one would expect a timeout if, for example, the token refresh URL wasn't resolvable. Either we are in an auth refresh loop, or the underlying aiohttp lib gets some corrupted response (we saw something similar in s3fs, but seems less likely here).

**Comment by yiga2** [comment]
> I would also try inverting the order
> 
> ```
> gcs = gcsfs.GCSFileSystem()
> gcs._connect_cloud()
> gcs.maybe_refresh()
> ```
> 
> but I suspect that
> 
> * creating the context (before calling the function, and after import, apparently) in the google lib creates additional information relevant to auth
> * the timing of creating the credentials object versus refreshing the token is not the same as it used to be, but I'm not sure what the difference is.
> 
> Thanks for investigating and bearing with us...

Same result - still hanging.

**Comment by yiga2** [comment]
> That still doesn't answer why the calls hang indefinitely - one would expect a timeout if, for example, the token refresh URL wasn't resolvable. Either we are in an auth refresh loop, or the underlying aiohttp lib gets some corrupted response (we saw something similar in s3fs, but seems less likely here).

Would `aiohttp` be called if `asynchronous=False` (the default) ?

**Comment by martindurant** [comment]
> Would aiohttp be called if asynchronous=False (the default) ?

Yes, it is the only backend. `asynchronous` indicated whether you intend to call the `aync def` functions directly, because the calling code is already running within an event loop. If False, we still use a loop internally to allow, for example, fetching and putting many files concurrently.

**Comment by yiga2** [comment]
Cleaning up - not sure what release fixed it but could not reproduce with 2021.06
